### ISSUE TITLE
Bug: Infinite iterator in props causes crash in react-dev-tools

### ISSUE THREAD

**Issue by mizukami234** [facebook/react#19660]
React version: Any
React Developer Tools: 4.8.2

## Steps To Reproduce

1. create and render a Component which contains infinite iterator

```js
function* makeCounter() {
  let c = 0;
  while (true) {
    yield c++;
  }
}
const counter = makeCounter();

const Sample = () => <Component counter={counter} />;
```

2. open chrome devtools and select this component with React Dev Tools (Components tab)

3. crashes after few seconds

Link to code example:

https://codesandbox.io/s/happy-lake-c1vyf?file=/src/App.js

## The current behavior

The page crashes due to devtools crash.

## The expected behavior

devtools doesn't crash.

This is because ReactDevTools is trying to display elements inside iterator object with transforming it into array here: https://github.com/facebook/react/blob/b6e1d086043a801682ff01b00c7a623d529b46c0/packages/react-devtools-shared/src/utils.js#L606

Iterators could be generally infinite (and also have side effect, like counter in the example). So I think it should not be extracted to array here.

**Comment by eps1lon** [comment]
The code you described is not the same as in your codesandbox. Your codesandbox is problematic as is since its render function has a side-effect:

```jsx
export default class Sample extends React.Component {
  render() {
    const counter = makeCounter();
    //              ^^^^^^^^^^^^^ every time render runs (with the same input) the output is different
    return <GeneratorInProps counter={counter} />;
  }
}
```

Devtools might work under the assumption that render is side-effect free and thus crash.

The code you described works fine: https://codesandbox.io/s/dawn-cdn-gf1m0?file=/src/App.js. So this might just be an issue that class components are treated differently.

The question is if devtools should not handle this and simply crash or guard against render functions with infinite iterators.

**Comment by mizukami234** [comment]
Thanks, @eps1lon .

Sorry, side-effect here was my misleading. I had to say that extracting iterator with `Array.from` to display props in devtools could cause unexpected crash with infinite iterator.

> The question is if devtools should not handle this and simply crash or guard against render functions with infinite iterators.

Yes, this is my point. I think there are 3 choices:

1. crash, no protection (current behavior)
2. limit the number of extracted items of iterator
3. don't extract iterator (most protective)

In my case, current behavior was really unexpected because I was simply passing a counter (infinite iterator) to props, and just counting inside a handler in a descendant component. 2 or 3 behavior is helpful for me.

**Comment by eps1lon** [comment]
Yeah an infinite iterator is not inherently bad. You could use it safely by calling next() in an effect.   The line you mentioned should probably iterate incrementally until some big number and then stop. 

Since you already identified the line of code would you mind opening a PR? 

Could you check if it's even safe for devtools to iterate over an iterator? wouldn't this change the behavior of the components using since devtools already moved the pointer ahead?

**Comment by omarsy** [comment]
I have a fix but it’s related to what @gaearon says in https://github.com/facebook/react/issues/19726 . I don’t know if I can publish it?

**Comment by eps1lon** [comment]
> I have a fix but it’s related to what @gaearon says in #19726 . I don’t know if I can publish it?

That issue has already someone working on it. Please be so kind and wait with the PR. If the assigned person stops working on it you can go ahead and propose your change.

**Comment by omarsy** [comment]
Ok I’ll wait  @todortotev before I publish my PR.

**Comment by engprodigy** [comment]
is the solution to this now available?

**Comment by engprodigy** [comment]
```
function* makeFiniteCounter(n) {
  let c = 0;
  while (c < n) {
    yield c++;
  }
}

if (makeFiniteCounter().toString() === '[object Generator]') {

       do something .............
}
```
Knowing that it is  a generator might allow you to differentiate from an iterable and process accordingly

**Comment by bvaughn** [comment]
This should be resolved by #19831 once it lands.
